hyper-capi/src/task.rs

```diff
@@ -0,0 +1,345 @@
+            future: Box::pin(async move { fut.await.into_dyn_task_type() }),
+            output: None,
+            userdata: AssertSendSafe(ptr::null_mut()),
+        })
+    }
+
+    fn output_type(&self) -> TaskType {
+        match self.output {
+            None => TaskType::Empty,
+            Some(ref val) => val.as_task_type(),
+        }
+    }
+}
+
+impl Future for TaskFuture {
+    type Output = Box<Task>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match Pin::new(&mut self.task.as_mut().unwrap().future).poll(cx) {
+            Poll::Ready(val) => {
+                let mut task = self.task.take().unwrap();
+                task.output = Some(val);
+                Poll::Ready(task)
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+ffi_fn! {
+    fn hyper_task_free(task: *mut Task) {
+        drop(unsafe { Box::from_raw(task) });
+    }
+}
+
+ffi_fn! {
+    fn hyper_task_value(task: *mut Task) -> *mut c_void {
+        if task.is_null() {
+            return ptr::null_mut();
+        }
+
+        let task = unsafe { &mut *task };
+
+        if let Some(val) = task.output.take() {
+            Box::into_raw(val) as *mut c_void
+        } else {
+            ptr::null_mut()
+        }
+    }
+}
+
+ffi_fn! {
+    fn hyper_task_type(task: *mut Task) -> TaskType {
+        if task.is_null() {
+            // instead of blowing up spectacularly, just say this null task
+            // doesn't have a value to retrieve.
+            return TaskType::Empty;
+        }
+
+        unsafe { &*task }.output_type()
+    }
+}
+
+ffi_fn! {
+    fn hyper_task_set_data(task: *mut Task, userdata: *mut c_void) {
+        if task.is_null() {
+            return;
+        }
+
+        unsafe { (*task).userdata = AssertSendSafe(userdata) };
+    }
+}
+
+ffi_fn! {
+    fn hyper_task_userdata(task: *mut Task) -> *mut c_void {
+        if task.is_null() {
+            return ptr::null_mut();
+        }
+
+        unsafe { &*task }.userdata.0
+    }
+}
+
+// ===== impl AsTaskType =====
+
+unsafe impl AsTaskType for () {
+    fn as_task_type(&self) -> TaskType {
+        TaskType::Empty
+    }
+}
+
+unsafe impl AsTaskType for hyper::Error {
+    fn as_task_type(&self) -> TaskType {
+        TaskType::Error
+    }
+}
+
+impl<T> IntoDynTaskType for T
+where
+    T: AsTaskType + Send + Sync + 'static,
+{
+    fn into_dyn_task_type(self) -> BoxAny {
+        Box::new(self)
+    }
+}
+
+impl<T> IntoDynTaskType for hyper::Result<T>
+where
+    T: AsTaskType + Send + Sync + 'static,
+{
+    fn into_dyn_task_type(self) -> BoxAny {
+        match self {
+            Ok(val) => Box::new(val),
+            Err(err) => Box::new(err), //TaskType::Error,
+        }
+    }
+}
+
+// ===== impl Waker =====
+
+
+ffi_fn! {
+    fn hyper_context_waker(cx: *mut Context<'_>) -> *mut Waker {
+        let waker = unsafe { &mut *cx }.waker().clone();
+        Box::into_raw(Box::new(Waker { waker }))
```

Comment:
I note that this requires allocation on every waker construction. Would it be better to create a two-pointer-wide struct on the C side to hold the waker directly and allow us to avoid this extra allocation?

Reply:
It is unfortunate, but I hoped not too badly. I'm unsure if the exact size of `std::task::Waker` can be relied upon... It does say that it is only composed of the data pointer and the vtable pointer, but it doesn't explicitly guarantee that won't ever change...

Reply:
I've brought this up to [wg-async-foundation](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/size_of.3A.3A.3CWaker.3E()).